numpy/core/src/multiarray/descriptor.c

```diff
@@ -2797,9 +2797,56 @@ arraydescr_setstate(PyArray_Descr *self, PyObject *args)
                 }
             }
 #else
-            PyErr_Format(PyExc_ValueError,
-                "non-string names in Numpy dtype unpickling");
-            return NULL;
+            // This section is an adaption of the block above (which reads PY2 pickles from PY3)
+            // to enable reading PY3 pickles from PY2. See gh-2407
+            PyObject *tmp, *new_name, *field;
+
+            tmp = PyDict_New();
+            if (tmp == NULL) {
+                return NULL;
+            }
+            Py_XDECREF(self->fields);
+            self->fields = tmp;
+
+            tmp = PyTuple_New(PyTuple_GET_SIZE(names));
+            if (tmp == NULL) {
+                return NULL;
+            }
+            Py_XDECREF(self->names);
+            self->names = tmp;
+
+            for (i = 0; i < PyTuple_GET_SIZE(names); ++i) {
+                name = PyTuple_GET_ITEM(names, i);
+                field = PyDict_GetItem(fields, name);
+                if (!field) {
+                    return NULL;
+                }
+
```

Comment:
The code block 2825 - 2843 is the only part of this that changes between python3 and python2. Please refactor to reduce the repeated code.